integration_test/integration_test.go

```diff
@@ -293,9 +359,643 @@ func (t *WASMIntegrationTestSuite) Test04_ContractExecution() {
+	)
+
+	fmt.Println("Pay very low fee. Should fail")
+	feeAmt := sdktypes.NewDec(xplaLowGasLimit).Mul(sdktypes.MustNewDecFromStr(xplaGasPrice))
+	fee := sdktypes.Coin{
+		Denom:  "axpla", // ordinal native fee
+		Amount: feeAmt.Ceil().RoundInt(),
+	}
+
+	txhash, _ := t.ValidatorWallet1.SendTx(ChainID, trasnferMsg, fee, xplaLowGasLimit, false)
+
+	err := txCheck(txhash)
+	if assert.Error(t.T(), err) {
+		fmt.Println("Error confirmed after broadcasted:", err)
+	} else {
+		fmt.Println("No error detected. Test fail")
+		t.T().Fail()
+	}
+}
+
+func (t *WASMIntegrationTestSuite) Test08_TxFeePaidByCw20_NativeTransfer_LackOfFee_SimulationShouldFail() {
+	fmt.Println("Preparing transfer message")
+
+	trasnferMsg := banktype.NewMsgSend(
+		t.ValidatorWallet1.ByteAddress, t.ValidatorWallet2.ByteAddress,
+		sdktypes.NewCoins(sdktypes.NewCoin("axpla", sdktypes.NewInt(1_000000_000000_000000))),
+	)
+
+	fmt.Println("Pay very low fee. Should fail")
+	feeAmt := sdktypes.NewDec(xplaLowGasLimit).Mul(sdktypes.MustNewDecFromStr(tknGasPrice))
+	fee := sdktypes.Coin{
+		Denom:  "uTKN", // use XATP
+		Amount: feeAmt.Ceil().RoundInt(),
+	}
+
+	txhash, _ := t.ValidatorWallet1.SendTx(ChainID, trasnferMsg, fee, xplaLowGasLimit, false)
+
+	err := txCheck(txhash)
+	if assert.Error(t.T(), err) {
+		fmt.Println("Error confirmed after broadcasted:", err)
+	} else {
+		fmt.Println("No error detected. Test fail")
+		t.T().Fail()
+	}
+}
+
+func (t *WASMIntegrationTestSuite) Test09_TxFeePaidByCw20_NativeTransfer() {
+	fmt.Println("Preparing transfer message")
+
+	trasnferMsg := banktype.NewMsgSend(
+		t.ValidatorWallet1.ByteAddress, t.ValidatorWallet2.ByteAddress,
+		sdktypes.NewCoins(sdktypes.NewCoin("axpla", sdktypes.NewInt(1_000000_000000_000000))),
+	)
+
+	fmt.Println("Pay fee as 5.1 TKN (= 0.51 XPLA)")
+	feeAmt := sdktypes.NewDec(xplaXATPTransferGasLimit).Mul(sdktypes.MustNewDecFromStr(tknGasPrice))
+	fee := sdktypes.Coin{
+		Denom:  "uTKN",
+		Amount: feeAmt.Ceil().RoundInt(),
+	}
+
+	txhash, err := t.ValidatorWallet1.SendTx(ChainID, trasnferMsg, fee, xplaXATPTransferGasLimit, false)
+	if assert.NotEqual(t.T(), "", txhash) && assert.NoError(t.T(), err) {
+		fmt.Println("Tx sent", txhash)
+	} else {
+		fmt.Println(err)
+	}
+
+	err = txCheck(txhash)
+	if assert.NoError(t.T(), err) {
+		fmt.Println("Tx applied", txhash)
+	} else {
+		fmt.Println(err)
+	}
+}
+
+func (t *WASMIntegrationTestSuite) Test10_TxFeePaidByCw20_Cw20Transfer() {
+	fmt.Println("Preparing contract execution message")
+
+	transferMsg := []byte(fmt.Sprintf(`
+		{
+			"transfer": {
+				"recipient": "%s",
+				"amount": "100000000"
+			}
+		}
+	`, t.UserWallet1.StringAddress))
+
+	contractExecMsg := &wasmtype.MsgExecuteContract{
+		Sender:   t.UserWallet2.StringAddress,
+		Contract: t.TokenAddress,
+		Msg:      transferMsg,
+	}
+
+	fmt.Println("Pay fee as 5.1 TKN (= 0.51 XPLA)")
+	feeAmt := sdktypes.NewDec(xplaXATPTransferGasLimit).Mul(sdktypes.MustNewDecFromStr(tknGasPrice))
```

Comment:
It fails if we enter the correct gas_price. I think we need 2 test cases. 
The first same gas price will fail. Second it reflects tax price, it will succeed

Reply:
https://github.com/xpladev/xpla/pull/60/commits/3d58e61983b11b13a91450924cc22f670d2669d8